Taint body on invalid call ABI

compiler-errors opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/142969

I'm not certain if there are any other paths that should be tainted, but they would operate similarly. Perhaps pointer coercion.

Introduces `extern "rust-invalid"` for testing purposes.

r? @workingjubilee or @oli-obk (or anyone)

compiler-errors committed:
Taint body on invalid call ABI

compiler-errors commented on tests/ui/abi/invalid-call-abi.rs:
```diff
@@ -0,0 +1,14 @@
+
+#![feature(abi_c_cmse_nonsecure_call)]
+
+const extern "C-cmse-nonsecure-call" fn foo() {
```
I am unfortunately unaware of the internals of this ABI, but do I need to limit this test's targets so that it doesn't pass where it shouldn't?

compiler-errors replied:
Also, I said this in zulip but I'd really like `extern "rustc-invalid"` which is an always-invalid ABI for testing purposes, for things just like this.

workingjubilee commented:
I believe we can coerce a function pointer from `unsafe extern "abi" fn(A) -> R` to `extern "abi" fn(A) -> R`.

workingjubilee commented on tests/ui/abi/invalid-call-abi.rs:
```diff
@@ -0,0 +1,14 @@
+// Fix for #142969.
```
```suggestion
// Fix for #142969 where an invalid ABI in a signature still had its call ABI computed
// because CTFE tried to evaluate it, despite previous errors during AST-to-HIR lowering.
```

compiler-errors committed:
Add rust-invalid ABI

compiler-errors commented:
Added `extern "rust-invalid"` in the first commit, and (manually...) checked that it still ICEd in the same way as `extern "cmse-*"`. Then I fixed the ICE in the second commit.

workingjubilee commented:
neat!

workingjubilee commented on tests/ui/abi/invalid-call-abi-ctfe.rs:
hmm, seems there's two of this test now? or perhaps I'm not seeing where the differences matter

workingjubilee commented on tests/ui/abi/invalid-call-abi-ctfe.rs:
```diff
@@ -0,0 +1,15 @@
+// Fix for #142969 where an invalid ABI in a signature still had its call ABI computed
+// because CTFE tried to evaluate it, despite previous errors during AST-to-HIR lowering.
+
+#![feature(abi_c_cmse_nonsecure_call)]
+
+const extern "C-cmse-nonsecure-call" fn foo() {
```
```suggestion
#![feature(rustc_attrs)]

const extern "rust-invalid" fn foo() {
```

compiler-errors replied:
lmfao

workingjubilee commented on tests/ui/abi/invalid-call-abi-ctfe.rs:
```diff
@@ -0,0 +1,15 @@
+#![feature(abi_c_cmse_nonsecure_call)]
+
+const extern "C-cmse-nonsecure-call" fn foo() {
+    //~^ ERROR "C-cmse-nonsecure-call" is not a supported ABI for the current target
+    //~| ERROR the `"C-cmse-nonsecure-call"` ABI is only allowed on function pointers
```
```suggestion
    //~^ ERROR "invalid" is not a supported ABI for the current target
```

workingjubilee reviewed (approved):
r=me with the apparent duplication sorted or explained

workingjubilee commented:
@bors r+ rollup

bors commented:
:pushpin: Commit e776065164f22872e8cadf5bc5e47352c27982dc has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee e776065164f22872e8cadf5bc5e47352c27982dc -->
<!-- homu: {"type":"Approved","sha":"e776065164f22872e8cadf5bc5e47352c27982dc","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

apiraino commented:
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202025-07-17/near/529287465). A backport PR will be authored by the release team at the end of the current development cycle. Backport labels handled by them.

@rustbot label +beta-accepted
